### ISSUE TITLE
Add multiple exclusion areas for bed clips

### ISSUE THREAD

**Issue by knightslaw** [SoftFever/OrcaSlicer#5922]
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

I upgraded to a G10 build plate that is held down with clips. My build space is small 180x180. So being able to accurately exclude 4 different areas would make love so much better!

### Which printers will be beneficial to this feature?

Marlin, All

### Describe the solution you'd like

The solution should be able to mark off small sections on the build plate where the clips are in the way. Slicer understands where these areas are and be able to route the nozzle around these areas . Only the small areas were the clips are will be avoided. The space in-between the clips can still be used as the not waste build space

### Additional context

![PXL_20240629_075600094](https://github.com/SoftFever/OrcaSlicer/assets/167050521/9d80083f-a188-4e38-8c97-b685a3c202f4)

**Comment by foreachthing** [comment]
You could model your build plate (3D, stl) and use that as your "real" build plate. I've the same for our Ultimaker 2, which has these clips as well. BUT, arrange could have issues with a non-rectangular build plate.
Or, you can draw an svg and use this image on a rectangular build plate and manage these areas manually. That way arrange will work better. That's what I use, currently.

**Comment by tlhintoq** [comment]
Or make a PNG the size of the print bed and put your clips marked out on it, then make that image the texture for the printer.
That way it always comes up automatically as part of the printer definition.
I did that to add the printer model here so there was just no way I could slice for the wrong machine.
![image](https://github.com/SoftFever/OrcaSlicer/assets/11351506/8f374dfc-6d54-4fbf-9e25-005a2cd9d4f3)

**Comment by oldcrazyeye** [comment]
The problem with just adding an svg,png ect. is the toolhead will still travel across the masked zones which will cause collisions still if you need mask off for structural collisions like toolchangers and awd configurations

**Comment by zenduin** [comment]
This is already possible, you just need to learn how to express it as points of a polygon.
Here is my crazy example:
![image](https://github.com/user-attachments/assets/5580017c-cf57-4a94-b62e-cefad8a88cb3)
0x0, 12x0, 12x30, 2x30, 2x54, 0x54, 0x140, 234x140, 234x146, 230x146, 230x140, 0x140, 0x250, 20x250, 20x305, 190x305, 190x345, 275x345, 275x330, 350x330, 350x350, 0x350

Here is an example of 30x20mm exclusion zones at the corners of a 200x200mm build plate:
![image](https://github.com/user-attachments/assets/0b0ad644-ea24-4123-810f-453361a981b5)

0x0, 30x0, 30x20, 0x20, 0x180, 30x180, 30x200, 170x200, 170x180, 200x180, 200x20, 170x20, 170x0, 200x0, 200x200, 0x200

**Comment by pterodaktil02** [comment]
I make  zone as  described 

0x0, 10x0, 10x300, 290x300, 290x0, 300x0, 300x300, 0x300  
 
 
 but  when i add   model  and say fill  the bed   it say that  its to close  to exclusion area
![Screenshot_2](https://github.com/user-attachments/assets/3f6f82a4-3a9a-4e49-84ad-09015126e928)

**Comment by knightslaw** [comment]
@zenduin This was great Thank you!! the only thing is I am also getting the error @pterodaktil02 is getting.
![image](https://github.com/user-attachments/assets/546f7e14-2e40-4d7d-a120-050cabe67bbd)

**Comment by EchotreePrints** [comment]
Greetings Programs!
I found this issue after modifying my K1 to have a heatsink attached to the extruder motor, and was looking for ways to make more than one exclusion zone in Orca Slicer. @zenduin made a great post to generate multiple zones, but it seems as if the solution isn't perfect.
    
If I outline 4 squares, 20mm in length from each corner on each side, it seems like the slicer actually excludes 20mm from the entire side of the build plate.

The coordinates used: `0x0, 0x20, 20x20, 20x0, 200x0, 200x20, 220x20, 220x200, 200x200, 200x220, 20x220, 20x200, 0x200, 0x220, 220x220, 0x220` This should start a square area at 0x0, to 20x20, then follow an awkward path around each of the squares on the inside of the plate, before closing off the polygon on the outside, see image for clarification.
![Untitled - OrcaSlicer 12_8_2024 1_32_40 PM](https://github.com/user-attachments/assets/de8ab5c4-7b93-439b-bfda-dd27047c8ca9)
This creates what looks like a backwards exclusion zone, where I can only print where I don't want to, but if you look, you can see that it's not throwing an error, and will let me slice the plate, and if I move an object into the places I don't want to print, it throws the error like expected. (weird)

![Untitled - OrcaSlicer 12_8_2024 1_53_36 PM](https://github.com/user-attachments/assets/602780a0-b5b7-4c30-98d0-e4c63c1ce483)
![Captures - File Explorer 12_8_2024 1_54_37 PM](https://github.com/user-attachments/assets/e6e78a32-90a3-49e1-947c-3cf8add4cf81)

What frustrates me is that even with this awkward geometry, it seems to exclude the entire area around the center anyways, I can't slice an object that is less than 20mm from the edge, regardless of if it's in an exclusion zone or not.

![Untitled - OrcaSlicer 12_8_2024 1_49_45 PM](https://github.com/user-attachments/assets/5f7eef1d-e18d-4320-a9aa-b5c4a8a55fd6)
![Untitled - OrcaSlicer 12_8_2024 1_49_49 PM](https://github.com/user-attachments/assets/04668126-c382-4a6b-a677-97ad886b9859)
![Captures - File Explorer 12_8_2024 1_58_17 PM](https://github.com/user-attachments/assets/870db60e-e87f-451e-98b5-cbecc9dc6eb8)

I even tried making the exclusion have a non-zero width edge to see if that would fix it, but it doesn't change anything. The slicer still acts as if the entire edge is excluded.
![Untitled - OrcaSlicer 12_8_2024 2_08_27 PM](https://github.com/user-attachments/assets/2cfa08ce-1a36-4f49-85be-36bb1dd82d5c)


Maybe I missed something, but I absolutely second the request for multiple independent exclusion zones. This would make setting them much easier.

**Comment by dhenry437** [comment]
Anyone able to assist with creating the following? I wish to give the bed a "padding" of 10mm.

![Untitled Diagram](https://github.com/user-attachments/assets/100bab03-39a4-42e1-8059-35b262829ba6)

Everything I've done i just end up with triangles everywhere. Or is there a better way to do this in Orca?

Thanks

**Comment by gilly991** [comment]
It's already enabled in the toolset!! ... you just need to create a polygon using cartesian coordinates, ie in a single direction avoiding crossing the polygon sides or origin point.  Here's my old Ender 3 V2, with std glass bed 235mm x 235mm plate, origin set at 3,3 and all around exclusion zones to suit ~8mm wide bed clips in Y and 1mm each side in X to clear the std purge line & just land on bed etc.  This gives an actual max print area of 224 x 219mm.
![image](https://github.com/user-attachments/assets/d47eb3b1-5fb3-4df2-9201-1a0500c28dbc)
![image](https://github.com/user-attachments/assets/751e4e8c-5586-4809-81c7-61348c37b40c)

**Comment by KopiasCsaba** [comment]
Yeah, it seems to be impossible, even if you get the drawing right, it will warn about "Cube is too close to exclusion area, there may be collisions when printing.

I have attached a 3mf file, hopefully containing everything. But, the key is the exclusion area (build plate 200x200+0+0).
0x0, 6x0, 6x90, 0x90, 0x190, 38x190, 38x200, 178x200, 178x190, 200x190, 200x90, 197x90, 197x0, 200x0, 200x200, 0x200, 0x0

And i have placed little cubes around the exclusion zones.

[Cube.3mf.txt](https://github.com/user-attachments/files/19521346/Cube.3mf.txt) (rename from .txt to .3mf)

For now i have drawn it as an svg, but it does not prevent the slicer to make a move that crosses the boundary, so it is really not a solution as I need to be careful still.

**Comment by realjackwang** [comment]
> Anyone able to assist with creating the following? I wish to give the bed a "padding" of 10mm.
> 
> ![Untitled Diagram](https://github.com/user-attachments/assets/100bab03-39a4-42e1-8059-35b262829ba6)
> 
> Everything I've done i just end up with triangles everywhere. Or is there a better way to do this in Orca?
> 
> Thanks

Here you are, 
```
0x0, 220x0, 220x220, 0x220, 0x0, 10x0, 10x210, 210x210, 210x10, 10x10, 10x0
```

It looks like the exclude bed area can only be a contiguous area and can't be set for each side individually, the padding code is generated by **ChatGPT**

**Comment by Nuck-TH** [comment]
This very much needs reimplementation, because existing way isn't just very unintuitive to add zones, they can't be reliably added on opposing sides of bed, where clips are usually placed.

**Comment by ForbiddenEra** [comment]
> This very much needs reimplementation, because existing way isn't just very unintuitive to add zones, they can't be reliably added on opposing sides of bed, where clips are usually placed.

Yeah it's not very intuitive, I mean, it's just vertexes for polygons but not actually having multiples and having to 'jank' it to make it work, annoying..

Close for Ender 3 V3 SE/KE bed screws and purge area but not quite perfect yet.. heh

<img width="1097" height="1048" alt="Image" src="https://github.com/user-attachments/assets/f7da49d2-2329-4a1b-ac63-6820c485d5af" />

**Comment by gilly991** [comment]
Bump.  My post above is still relevant and the coordinates give the desired exclusion zones on my bed, but after a firmware update on my Ender 3v2 (mrsicoc) it started throwing an error every time.  This orca coding solution must have some glitches or conflicts with general Marlin flavour firmware (tried changing flavour in the slicer but appears no different) or maybe just the slicer itself.  Could do with sorting, as seems basic functionality request.

**Comment by TheInnoventor** [comment]
@gilly991 @EchotreePrints @KopiasCsaba  What fixed this for me was putting a sort of notch down in the bottom left rectangle to allow space for the prime line (shown in red). Orca wasn't showing it (because of the error preventing slicing at all) but the bottom end of the prime line was overlapping with the bottom left rectangle I excluded for a bed clip. Now my 4 separate rectangular exclusion zones seems to be working.

<img width="1730" height="1030" alt="Image" src="https://github.com/user-attachments/assets/431bf059-e450-4244-b6fb-ba036f8fcc73" />

**Comment by stahlsau** [comment]
this still doesn't seem to work correctly - or maybe I got it totally wrong, I dunno.

<img width="1144" height="796" alt="Image" src="https://github.com/user-attachments/assets/b68e88e4-a7b5-4656-a932-ee8d103dd689" />
I drew the outline and then the corners:
0x0, 300x0, 300x300, 0x300, 0x30, 30x0, 270x0, 300x30, 300x270, 270x300, 30x300, 0x270

The drawing seems fine but I can't slice anything.

**Comment by stahlsau** [comment]
direct answer to my own post - if I go the other way around, it works: first the outer bounds counter-clockwise, then the inner bounds clockwise. No idea why.

<img width="847" height="623" alt="Image" src="https://github.com/user-attachments/assets/e1bfecbb-6928-4e5b-8f48-4304263adec6" />

0x0, 300x0, 300x300, 0x300, 0x270, 30x300, 270x300, 300x270, 300x30, 270x0, 30x0, 0x30

**Comment by Harry-Loud** [comment]
Hey guys, see #13777 for our implementation that gives you multiple exclusion zones and 3D exclusion zones.
It should hopefully be merged into Orca soon.

Harry Loud
Muon3D
